File: game/quest.py
from rich.panel import Panel

from game.utils.ui import c
from game.data import stats


class Quest:
    def __init__(self) -> None:
        self.stats = stats

    def show_quest(self) -> None:
        quest = f"""
        I. Combat
        {self.defeat_enemy()}

        II. Mining
        {self.mine_stone()}

        III. Gourmet
        Eat 10 Foods
        """
        c.print(Panel(quest, title="Active Quest", expand=False))

    def defeat_enemy(self) -> str:
        if self.stats["enemy_defeated"] in range(0, 9):
            return "[italic green]Defeat 10 Enemy[/italic green]"
        elif self.stats["enemy_defeated"] in range(10, 19):
            return "[italic green]Defeat 20 Enemy[/italic green]"
        elif self.stats["enemy_defeated"] in range(20, 29):
            return "[italic green]Defeat 50 Enemy[/italic green]"
        elif self.stats["enemy_defeated"] in range(30, 99):
            return "[bold green]Defeat 100 Enemy[/bold green]"
        elif self.stats["enemy_defeated"] in range(100, 999999999):
            return f"""
        [bold magenta]Quest Combat Complete![/bold magenta]
        [italic yellow]Enemy Defeated: {self.stats["enemy_defeated"]}[/italic yellow]
        """

    def mine_stone(self) -> str:
        if self.stats["stone_collected"] in range(0, 19):
            return "[italic blue]Mine 20 Stone[/italic blue]"
        elif self.stats["stone_collected"] in range(20, 39):
            return "[italic blue]Mine 40 Stone[/italic blue]"
        elif self.stats["stone_collected"] in range(40, 59):
            return "[italic blue]Mine 60 Stone[/italic blue]"
        elif self.stats["stone_collected"] in range(60, 79):
            return "[italic blue]Mine 80 Stone[/italic blue]"
        elif self.stats["stone_collected"] in range(80, 99):
            return "[italic blue]Mine 100 Stone[/italic blue]"
        elif self.stats["stone_collected"] in range(100, 999999999):
            return f"""
        [bold magenta]Quest Mining Complete![/bold magenta]
        [italic yellow]Stone Collected: {self.stats["stone_collected"]}[/italic yellow]
        """

    def to_dict(self) -> None:
        return self.stats

    def from_dict(self, data) -> None:
        self.stats = data

File: src/game/quest.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use crate::game::data::{quests, initial_stats};

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct Quest {
    pub stats: HashMap<String, i32>,
}

impl Quest {
    pub fn new() -> Self {
        Self {
            stats: initial_stats(),
        }
    }

    pub fn show_quest(&self) {
        println!("Active Quests:");
        let q = quests();

        println!("I. Combat");
        println!("{}", self.quest_progress("enemy_defeated", "Defeat", &q["combat"], "Enemy", "green"));

        println!("II. Mining");
        println!("{}", self.quest_progress("stone_collected", "Mine", &q["mining"], "Stone", "blue"));
    }

    fn quest_progress(&self, key: &str, action: &str, milestones: &[i32], item: &str, _color: &str) -> String {
        let value = *self.stats.get(key).unwrap_or(&0);

        for &milestone in milestones {
            if value < milestone {
                return format!("{} {} {}", action, milestone, item);
            }
        }

        if value > 100 {
             return format!("Complete | {} {} {}", action, value, item);
        }

        format!("Quest {} Complete!\n{} Count: {}", action, action, value)
    }
}
